### PR TITLE
Add support for user banners

### PR DESCRIPTION
**Pull request by conorwastakenwastaken**
With banners being released, a lot of people have asked Discord to allow bots to see user banners, and they have [confirmed that they will add the banner image hash to the user object](https://github.com/discord/discord-api-docs/issues/3095#issuecomment-868817186) in the future. This PR also adds the `bannerColor` property to the User structure, which is also being exposed to us (if the user sets one)!

~~Due to the fact that this PR is untestable, and the property is not yet sent to bots, I'll mark it as a draft. This is currently only available in the Canary API.~~

These changes are documented [here](https://github.com/discord/discord-api-docs/pull/3448#issue-690275742).

### CODE REVIEW

File: lib/structures/User.js [hunk 1]
```diff
@@ -97,6 +109,15 @@ class User extends Base {
         return this.avatar ? this._client._formatImage(Endpoints.USER_AVATAR(this.id, this.avatar), format, size) : this.defaultAvatarURL;
     }
 
+    /**
+    * Get the user's banner with the given format and size
+    * @arg {String} [format] The filetype of the banner ("jpg", "jpeg", "png", "gif", or "webp")
+    * @arg {Number} [size] The size of the banner (any power of two between 16 and 4096)
+    */
+    dynamicBannerURL(format, size) {
```
Comment (bsian03) [inline lib/structures/User.js:132]: Return type required

File: index.d.ts [hunk 2]
```diff
@@ -2584,6 +2589,7 @@ declare namespace Eris {
     addRelationship(block?: boolean): Promise<void>;
     deleteNote(): Promise<void>;
     dynamicAvatarURL(format?: ImageFormat, size?: number): string;
+    dynamicBannerURL(format?: ImageFormat, size?: number): string;
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
    dynamicBannerURL(format?: ImageFormat, size?: number): string | null;
```
Can also return `null`

File: lib/structures/User.js [hunk 3]
```diff
@@ -53,6 +58,13 @@ class User extends Base {
         return this.avatar ? this._client._formatImage(Endpoints.USER_AVATAR(this.id, this.avatar)) : this.defaultAvatarURL;
     }
 
+    get bannerURL() {
+        if(this._missingClientError) {
```
Comment (bsian03) [inline lib/structures/User.js:69]: We can probably return null if there is no banner before checking if the client exists

**Review by bsian03 — CHANGES_REQUESTED** [review]

File: lib/structures/User.js (hunk 1 above)
Comment (conorwastakenwastaken) [reply lib/structures/User.js:132]: The return type is missing on all of the other dynamic URL functions, so I figured it wasn't needed. I'll go through and add it to all the other ones with this commit if you think it's necessary.

File: index.d.ts (hunk 2 above)
Comment (conorwastakenwastaken) [reply index.d.ts]: See [this comment](https://github.com/abalabahaha/eris/pull/1238#discussion_r663509170), I did the same thing for this file. 👍🏻

File: lib/structures/User.js [hunk 4]
```diff
@@ -92,11 +107,22 @@ class User extends Base {
...
     }
 
+    /**
+    * Get the user's banner with the given format and size
+    * @arg {String} [format] The filetype of the banner ("jpg", "jpeg", "png", "gif", or "webp")
+    * @arg {Number} [size] The size of the banner (any power of two between 16 and 4096)
+    * @returns {String?}
+    */
+    dynamicBannerURL(format, size) {
+        return this.banner ? this._client._formatImage(Endpoints.BANNER(this.id, this.banner), format, size) : null;
```
Comment (bsian03) [inline lib/structures/User.js]: This still needs a missing client checker

**Review by bsian03 — CHANGES_REQUESTED** [review]

**Comment by Awoocado** [comment]
Those fields are accessible in the canary API and documented in discord/discord-api-docs#3448

**Comment by conorwastakenwastaken** [comment]
Regarding a40c55a, please see [this comment](https://github.com/discord/discord-api-docs/pull/3448#issuecomment-881053324).

**Comment by conorwastakenwastaken** [comment]
`banner` and `banner_color` are in the stable API now, but it'd probably be best for us to wait until Discord update the API again to fully turn `banner_color` to int?

**Comment by GoldenAngel2** [comment]
It would be best to wait until they update it to int.

File: lib/gateway/Shard.js [hunk 5]
```diff
@@ -1948,7 +1954,9 @@ class Shard extends EventEmitter {
                 const oldUser = {
                     username: user.username,
                     discriminator: user.discriminator,
-                    avatar: user.avatar
+                    avatar: user.avatar,
+                    banner: user.banner,
+                    bannerColor: user.bannerColor
```
Comment (GoldenAngel2) [inline lib/gateway/Shard.js]: `bannerColor` -> `accentColor`

File: lib/gateway/Shard.js [hunk 6]
```diff
@@ -1108,11 +1112,13 @@ class Shard extends EventEmitter {
...
                     let oldUser = null;
-                    if(user && (user.username !== packet.d.user.username || user.avatar !== packet.d.user.avatar || user.discriminator !== packet.d.user.discriminator)) {
+                    if(user && (user.username !== packet.d.user.username || user.discriminator !== packet.d.user.discriminator || user.avatar !== packet.d.user.avatar || user.banner !== packet.d.user.banner || user.bannerColor !== packet.d.user.banner_color)) {
                         oldUser = {
                             username: user.username,
                             discriminator: user.discriminator,
-                            avatar: user.avatar
+                            avatar: user.avatar,
+                            banner: user.banner,
+                            bannerColor: user.bannerColor
```
Comment (GoldenAngel2) [inline lib/gateway/Shard.js]: `bannerColor` -> `accentColor`

File: lib/gateway/Shard.js [hunk 7]
```diff
@@ -622,6 +624,8 @@ class Shard extends EventEmitter {
                         * @prop {String} oldUser.username The username of the user
                         * @prop {String} oldUser.discriminator The discriminator of the user
                         * @prop {String?} oldUser.avatar The hash of the user's avatar, or null if no avatar
+                        * @prop {String?} oldUser.banner The hash of the user's banner, or null if no banner
+                        * @prop {Number?} oldUser.bannerColor The user's banner color, or null if no banner color
```
Comment (GoldenAngel2) [inline lib/gateway/Shard.js]: `bannerColor` -> `accentColor`

File: lib/gateway/Shard.js [hunk 8]
```diff
@@ -605,11 +605,13 @@ class Shard extends EventEmitter {
...
                     let oldUser = null;
-                    if(user && (user.username !== packet.d.user.username || user.avatar !== packet.d.user.avatar || user.discriminator !== packet.d.user.discriminator)) {
+                    if(user && (user.username !== packet.d.user.username || user.discriminator !== packet.d.user.discriminator || user.avatar !== packet.d.user.avatar || user.banner !== packet.d.user.banner || user.bannerColor !== packet.d.user.banner_color)) {
                         oldUser = {
                             username: user.username,
                             discriminator: user.discriminator,
-                            avatar: user.avatar
+                            avatar: user.avatar,
+                            banner: user.banner,
+                            bannerColor: user.bannerColor
```
Comment (GoldenAngel2) [inline lib/gateway/Shard.js]: `bannerColor` -> `accentColor`

File: lib/gateway/Shard.js [hunk 9]
```diff
@@ -605,11 +605,13 @@ class Shard extends EventEmitter {
                 if(packet.d.user.username !== undefined) {
                     let user = this.client.users.get(packet.d.user.id);
                     let oldUser = null;
-                    if(user && (user.username !== packet.d.user.username || user.avatar !== packet.d.user.avatar || user.discriminator !== packet.d.user.discriminator)) {
+                    if(user && (user.username !== packet.d.user.username || user.discriminator !== packet.d.user.discriminator || user.avatar !== packet.d.user.avatar || user.banner !== packet.d.user.banner || user.bannerColor !== packet.d.user.banner_color)) {
```
Comment (GoldenAngel2) [inline lib/gateway/Shard.js]: `bannerColor` -> `accentColor`
`banner_color` -> `accent_color`

File: index.d.ts [hunk 10]
```diff
@@ -2569,6 +2574,9 @@ declare namespace Eris {
   export class User extends Base {
     avatar: string | null;
     avatarURL: string;
+    banner: string | null;
+    bannerColor: number | null;
```
Comment (GoldenAngel2) [inline index.d.ts]: `bannerColor` -> `accentColor`

File: index.d.ts [hunk 11]
```diff
@@ -2193,6 +2195,9 @@ declare namespace Eris {
     activities?: Activity[];
     avatar: string | null;
     avatarURL: string;
+    banner: string | null;
+    bannerColor: number | null;
```
Comment (GoldenAngel2) [inline index.d.ts]: `bannerColor` -> `accentColor`

File: index.d.ts [hunk 12]
```diff
@@ -798,6 +798,8 @@ declare namespace Eris {
   }
   interface PartialUser {
     avatar: string | null;
+    banner: string | null;
+    bannerColor: number | null;
```
Comment (GoldenAngel2) [inline index.d.ts]: `bannerColor` -> `accentColor`

File: lib/gateway/Shard.js [hunk 13]
```diff
@@ -1108,11 +1112,13 @@ class Shard extends EventEmitter {
                 if(!(this.client.options.intents & Constants.Intents.guildPresences) && packet.d.user.username !== undefined) {
                     let user = this.client.users.get(packet.d.user.id);
                     let oldUser = null;
-                    if(user && (user.username !== packet.d.user.username || user.avatar !== packet.d.user.avatar || user.discriminator !== packet.d.user.discriminator)) {
+                    if(user && (user.username !== packet.d.user.username || user.discriminator !== packet.d.user.discriminator || user.avatar !== packet.d.user.avatar || user.banner !== packet.d.user.banner || user.bannerColor !== packet.d.user.banner_color)) {
```
Comment (GoldenAngel2) [inline lib/gateway/Shard.js]: `bannerColor` -> `accentColor`
`banner_color` -> `accent_color`

File: lib/structures/Member.js [hunk 14]
```diff
@@ -112,6 +115,18 @@ class Member extends Base {
         return this.user.avatarURL;
     }
 
+    get banner() {
+        return this.user.banner;
+    }
+
+    get bannerColor() {
+        return this.user.bannerColor;
```
Comment (GoldenAngel2) [inline lib/structures/Member.js]: `bannerColor` -> `accentColor`

File: lib/structures/User.js [hunk 15]
```diff
@@ -44,6 +47,12 @@ class User extends Base {
         if(data.public_flags !== undefined) {
             this.publicFlags = data.public_flags;
         }
+        if(data.banner !== undefined) {
+            this.banner = data.banner;
+        }
+        if(data.banner_color !== undefined) {
+            this.bannerColor = data.banner_color;
```
Comment (GoldenAngel2) [inline lib/structures/User.js]: `bannerColor` -> `accentColor`
`banner_color` -> `accent_color`

File: lib/structures/User.js [hunk 16]
```diff
@@ -133,6 +175,8 @@ class User extends Base {
     toJSON(props = []) {
         return super.toJSON([
             "avatar",
+            "banner",
+            "bannerColor",
```
Comment (GoldenAngel2) [inline lib/structures/User.js]: `bannerColor` -> `accentColor`

**Review by GoldenAngel2 — CHANGES_REQUESTED** [review]
`banner_color` has been changed to `accent_color`

`bannerColor` -> `accentColor`

https://github.com/discord/discord-api-docs/pull/3448

`banner_color` will most likely be removed whenever the old clients switch to `accent_color`

(also `accent_color` is the number/int)

**Comment by DonovanDMC** [comment]
Is it something with my setup, this pr, or are these not sent via gateway yet? I'm currently fetching the user via rest if the `banner`/`accent_color` properties are undefined, and I'm getting them just fine via rest.

(using [DonovanDMC/eris](https://github.com/DonovanDMC/eris/tree/everything) @ 5c00439680409a78b174144bbecc56059256a104)

**Comment by GoldenAngel2** [comment]
It's only available via rest. @DonovanDMC 
There's a feature request open to include it in the message author object https://github.com/discord/discord-api-docs/discussions/3577

**Comment by DonovanDMC** [comment]
that's annoying

File: index.d.ts [hunk 17]
```diff
@@ -797,7 +797,9 @@ declare namespace Eris {
     user: User;
   }
   interface PartialUser {
+    accentColor: number | null;
     avatar: string | null;
+    banner: string | null;
```
Comment (bsian03) [inline index.d.ts]: At the moment both of these are partial nullables as they don't seem to be coming through on gateway, only via REST

**Review by bsian03 — CHANGES_REQUESTED** [review]
I'll see if Discord are planning to change this but for now, make the change to reflect the current behaviour

**Comment by conorwastakenwastaken** [comment]
Messed up with 8200263 and ef632cf sorry lol

File: index.d.ts [hunk 18]
```diff
@@ -813,7 +813,9 @@ declare namespace Eris {
     roles: string[];
   }
   interface PartialUser {
+    accentColor?: number;
     avatar: string | null;
+    banner?: string;
```
Comment (bsian03) [inline index.d.ts]: They're still nullable, just optional too

**Review by bsian03 — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 18 above)
Comment (conorwastakenwastaken) [reply index.d.ts]: Mb

**Review by bsian03 — APPROVED** [review]

**Pull request merged**